### ISSUE TITLE
New snaphot - Error in dlopen: libtorch.so: cannot open shared object file

### ISSUE THREAD

**Issue by hmf** [sbrunk/storch#62]
While working on #61, I rebuilt my dev container. The build failed. Upon investigations I see that the old snapshot does not exist. We [now have](https://oss.sonatype.org/content/repositories/snapshots/org/bytedeco/cuda-platform-redist/):

[12.3-8.9-1.5.10-SNAPSHOT/](https://oss.sonatype.org/content/repositories/snapshots/org/bytedeco/cuda-platform-redist/12.3-8.9-1.5.10-SNAPSHOT/) 	Wed Oct 25 13:34:15 UTC 2023 	  	
[maven-metadata.xml](https://oss.sonatype.org/content/repositories/snapshots/org/bytedeco/cuda-platform-redist/maven-metadata.xml) 	Wed Oct 25 14:13:45 UTC 2023 	393 	
[maven-metadata.xml.md5](https://oss.sonatype.org/content/repositories/snapshots/org/bytedeco/cuda-platform-redist/maven-metadata.xml.md5) 	Wed Oct 25 14:13:45 UTC 2023 	32 	
[maven-metadata.xml.sha1](https://oss.sonatype.org/content/repositories/snapshots/org/bytedeco/cuda-platform-redist/maven-metadata.xml.sha1) 	Wed Oct 25 14:13:45 UTC 2023 	40 	
[maven-metadata.xml.sha256](https://oss.sonatype.org/content/repositories/snapshots/org/bytedeco/cuda-platform-redist/maven-metadata.xml.sha256) 	Wed Oct 25 14:13:45 UTC 2023 	64 	
[maven-metadata.xml.sha512](https://oss.sonatype.org/content/repositories/snapshots/org/bytedeco/cuda-platform-redist/maven-metadata.xml.sha512) 	Wed Oct 25 14:13:45 UTC 2023 	128

I changed the library (Ivy dependencies). Now the build is ok, but execution fails. I suspect the problem may be due to the use of the CUDA 12.3 version. I now get the following message when running an app:

```
vscode ➜ /workspaces/storch (explore_1) $ ./mill examples.runMain LeNetApp2
[109/116] examples.compile 
[info] compiling 1 Scala source to /workspaces/storch/out/examples/compile.dest/classes ...
[info] done compiling
[116/116] examples.runMain 
[W interface.cpp:47] Warning: Loading nvfuser library failed with: Error in dlopen: libtorch.so: cannot open shared object file: No such file or directory (function LoadingNvfuserLibrary)
Using device: Device(CUDA,-1)
Exception in thread "main" java.lang.ExceptionInInitializerError
        at LeNetApp2.main(LeNet2.scala)
Caused by: java.lang.RuntimeException: CUDA error: the provided PTX was compiled with an unsupported toolchain.
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

Exception raised from c10_cuda_check_implementation at /home/runner/work/javacpp-presets/javacpp-presets/pytorch/cppbuild/linux-x86_64-gpu/pytorch/c10/cuda/CUDAException.cpp:44 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x6c (0x7fc0e40b1d8c in /home/vscode/.javacpp/cache/pytorch-2.0.1-1.5.10-20231025.021128-64-linux-x86_64-gpu.jar/org/bytedeco/pytorch/linux-x86_64-gpu/libc10.so)
```

Can anyone confirm this is the most probable cause? Of course the issue now is installing the CUDA libraries in the dev-container. Only 12.2 is currently supported. Is installing 12.3 the only way forward?

TIA

**Comment by sbrunk** [comment]
Yes the native bindings have been updated to CUDA 12.3 as of https://github.com/bytedeco/javacpp-presets/commit/7d56a2cf9badfe9798a59358397f45f3f125894d. Could you check if you have a recent Nvidia driver installed? At least 535. I ran into this issue too with a 530 driver but after updating it started working again.

See also this [thread](https://forums.developer.nvidia.com/t/provided-ptx-was-compiled-with-an-unsupported-toolchain-error-using-cub/168292) and the [CUDA Compatibility docs](https://docs.nvidia.com/deploy/cuda-compatibility/).

There's some discussion about using upstream libtorch in https://github.com/bytedeco/javacpp-presets/pull/1426 which is compiled agains CUDA but it's unclear if this is feasible.

**Comment by hmf** [comment]
@sbrunk Thanks for the help. 

I have checked the compatibility matrix. Command `nvidia-smi`reports:

```
 Driver Version: 535.104.05   CUDA Version: 12.2 
```

`nvcc --version` reports:

```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Tue_Aug_15_22:02:13_PDT_2023
Cuda compilation tools, release 12.2, V12.2.140
Build cuda_12.2.r12.2/compiler.33191640_0
```

All seems to be good.  I also have:

```
vscode ➜ /workspaces/storch (explore_1) $ locate -i libtorch.so
/home/vscode/.javacpp/cache/pytorch-2.0.1-1.5.10-20231025.021128-64-linux-x86_64-gpu.jar/org/bytedeco/pytorch/linux-x86_64-gpu/libtorch.so
```

Could this be an issue of simply not finding `libtorch.so` as opposed to getting an incorrect version? Any other diagnostics I could try out?

**Comment by hmf** [comment]
Strange. I made an update & upgrade. Did a reboot. Now I get (libtorch warning):

```
vscode ➜ /workspaces/storch (explore_1) $ ./mill examples.runMain LeNetApp2
[116/116] examples.runMain 
Using device: Device(CUDA,-1)
Exception in thread "main" java.lang.ExceptionInInitializerError
        at LeNetApp2.main(LeNet2.scala)
Caused by: java.lang.RuntimeException: CUDA error: the provided PTX was compiled with an unsupported toolchain.
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

Exception raised from c10_cuda_check_implementation at /home/runner/work/javacpp-presets/javacpp-presets/pytorch/cppbuild/linux-x86_64-gpu/pytorch/c10/cuda/CUDAException.cpp:44 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x6c (0x7f9fe2a58d8c in /home/vscode/.javacpp/cache/pytorch-2.0.1-1.5.10-20231025.021128-64-linux-x86_64-gpu.jar/org/bytedeco/pytorch/linux-x86_64-gpu/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xfa (0x7f9fe2a1ef6a in /home/vscode/.javacpp/cache/pytorch-2.0.1-1.5.10-20231025.021128-64-linux-x86_64-gpu.jar/org/bytedeco/pytorch/linux-x86_64-gpu/libc10.so)
```

**Comment by sbrunk** [comment]
Hmm my assumption about the driver might have been wrong (it might be needed too, but not always enough).

While testing an update to PyTorch 2.1, I just compiled libtorch from source using CUDA 12.3. I tried to run some tests on the GPU then, and ran into the error you're seeing: `CUDA error: the provided PTX was compiled with an unsupported toolchain.`

I went back to main and the error still occurs. This is a machine with an A 6000 ADA GPU (ampere 86 arch).
On my laptop with an older GPU it's working fine though, so I suspect that the issue could be missing support for that `86` arch in our libtorch build in [javacpp-presets](https://github.com/bytedeco/javacpp-presets/blob/7d56a2cf9badfe9798a59358397f45f3f125894d/pytorch/cppbuild.sh#L30).

```
export TORCH_CUDA_ARCH_LIST="5.0;6.0;7.0+PTX"
```

@hmf what's the GPU you're testing this on?

**Comment by hmf** [comment]
@sbrunk  I have a NVIDIA GeForce RTX 3090 in the VM

**Comment by sbrunk** [comment]
According to https://developer.nvidia.com/cuda-gpus, the RTX 3090 has compute capability 8.6 just like the RTX A6000, supporting my assumption that we might need to update the `ARCH_LIST` of the libtorch build in JavaCPP.

@HGuillemet do you think this makes sense?

**Comment by sbrunk** [comment]
> EDIT: I assume you are saying that `TORCH_CUDA_ARCH_LIST="8.6"` should also be included. However, this did run before. I still have the 12.1-8.9 in the VM (not the dev container). So I ran that again. Tests on both `12.1`and `12.3` execute correctly.
> 
> Now I get he same error in the VM as the dev container for both `12.1`and `12.3` versions. I suspect some update might have messed this up.

Hmm yeah that's a bit weird. If I recall correctly I've seen the error too and then it did work again (after upgrading the driver) until I tested again today. Perhaps check your `~/.javacpp/cache` folder. In the pytorch folder there, some shared libs are symlinks to the cuda folder and I don't know if it updates them and how that mechanism interacts with snapshots.

**Comment by hmf** [comment]
I figured I would try to install CUDA Toolkit 12.3 in a Dev container just to see what happens. So I get this error when I try it out:

```
[2023-10-28T15:44:07.599Z] 59.40 Reading state information...
59.40 Package libcudnn8 is not available, but is referred to by another package.
59.40 This may mean that the package is missing, has been obsoleted, or
59.40 is only available from another source
59.40 
59.40 E: Version '8.9.5.29-1+cuda12.3' for 'libcudnn8' was not found
[2023-10-28T15:44:07.599Z] 59.40 ERROR: Feature "NVIDIA CUDA" (ghcr.io/devcontainers/features/nvidia-cuda) failed to install! Look at the documentation at https://github.com/devcontainers/features/tree/main/src/nvidia-cuda for help troubleshooting this error.
------
[2023-10-28T15:44:07.599Z] Dockerfile.extended:80
```
If I looked at the [cuDNN support matrix](https://docs.nvidia.com/deeplearning/cudnn/support-matrix/index.html), I see that the latest version for the [CUDA Toolkit Version](https://developer.nvidia.com/cuda-toolkit-archive) is `12.2`. 

I am assuming that libTorch compilation requires and links to CUDA and CuDNN. So how were the JavaCPP binaries created? In the build script I see variables pointing to a CuDNN installation. 

Maybe `12.3` is not ready fir use? Maybe step back for a `12.2` snapshot first?

**Comment by sbrunk** [comment]
I built libtorch from source against CUDA 12.3 and with added support for compute capability 8.0 and 8.6 but I'm still running into the same issue. I didn't check CuDNN before but now I'm seeing that my version is also for 12.2, not sure if this an issue though.

I'll try to compile libtorch against CUDA 12.1 next.

**Comment by hmf** [comment]
> https://github.com/bytedeco/javacpp-presets/blob/master/.github/actions/deploy-ubuntu/action.yml

So CuDNN version is `8.9.5.29-1.cuda12.2`, which is the the one I tested with CUDA 12.2. 

> I built libtorch from source against CUDA 12.3 and with added support for compute capability 8.0 and 8.6 

I have a hunch this might not be the issue. I think that if you don't add the capability, you simple don't take advantage of the hardware but it should still execute without problems. As I said, this seemed to work with  CUDA 11.8 (and respective CuDNN).  

> I'll try to compile libtorch against CUDA 12.1 next.

Hope that works. But then 12.2 is also a possible candidate, right? 

> For these tests, I'm using the Ubuntu CUDA/CuDNN packages whose installation is described 

Used the same instructions in the bare VM with the same negative results.

**Comment by HGuillemet** [comment]
I would also say that the "CUDA error: the provided PTX was compiled with an unsupported toolchain" is related to the CUDA driver version. But with version 535 it should not happen.

You can try running with libtorch binary downloadable from pytorch.org instead of the libraries in JavaCPP  (using `-Dorg.bytedeco.javacpp.pathsFirst`). It will not need PTX.

It's sure that we'd better avoid all problems related to PTX (this error, long startup time when compiling PTX, features lacking, size of cuda cache...) and provide cubins for all major architectures, but we are limited to only 1 or 2 architecture when compiling Pytorch presets because of the compilation time it requires and GitHub actions limits.

That's why I'd like to include the libtorch binaries provided by pytorch.org in JavaCPP presets. But I still need to test what is doable.

**Comment by sbrunk** [comment]
I've done some tests. I've put the results into the 2.1.0 update PR thread https://github.com/bytedeco/javacpp-presets/pull/1426#issuecomment-1783933140

**Comment by hmf** [comment]
@sbrunk Can you tell me what the status of this is? Bytedeco snapshot seem to be the same version. I have a CUDA driver 12.2 on the bare metal machine. Tests with Toolkit 11.8, 12.2, 12.1  and 12.3 on the dev container don't seem to work (they use the 12.2 driver). What is the plan for Storch? Move all CUDA to 12.3 (drivers 545.XX.YY)? At this moment, can I assume that the current snapshot won't woprk with the CUDA 12.2 + 535.XXX.YY drivers?

**Comment by sbrunk** [comment]
Yes but it should start working again once we update to PyTorch 2.1 https://github.com/sbrunk/storch/pull/64 which brings in fixes on the JavaCPP bindings. See https://github.com/bytedeco/javacpp-presets/pull/1426#issuecomment-1786998372.

The main reason why I haven't pushed the update yet is that PyTorch 2.1 adds new data types which we need to add as well.

**Comment by hmf** [comment]
@sbrunk Thanks for the information. 

> The main reason why I haven't pushed the update yet is that PyTorch 2.1 adds new data types which we need to add as well.

Hope that does not entail too many changes on my side.

**Comment by sbrunk** [comment]
I don't think so. The upgrade seems to be quite smooth so far. I've started working on integrating the datatypes and hopefully we'll be able to get it ready in the next few days.

**Comment by sbrunk** [comment]
Hey @hmf I finally had the bandwidth to get the update to PyTorch 2.1 https://github.com/sbrunk/storch/pull/64 done. Could you give it a try with the latest snapshot to check if the CUDA issues are solved for you?

**Comment by sbrunk** [comment]
@hmf just to be sure when trying to run with sbt did you enable GPU in the sbt build? I.e.
```
set ThisBuild / enableGPU := true
```
It defaults to false currently. If I set it to true, the build switches to GPU-enabled LibTorch and adds `cuda-redist` as dependency. With that, the `LeNet` example works on the machines I've tried so far.

The example checks `torch.cuda.isAvailable` and falls back to CPU otherwise.

**Comment by sbrunk** [comment]
`Using device: Device(CUDA,-1)` seems fine I get that too. In contrast to PyTorch LibTorch returns -1 when you don't explicitly enumerate a GPU I think.

Could you check the output of nvidia-smi or nvtop to see if it is actually using the GPU?

**Comment by hmf** [comment]
@sbrunk You are correct. `nvidia-smi` shows the GPU is being used. I will merge this into my latest and see how that goes. Thank you.
